### PR TITLE
CC | add sealed secret support in Kata

### PR DESCRIPTION
**Pull request by LindaYu17**
only support the secret as env yet

### CODE REVIEW

**Comment by LindaYu17** [comment]
@Xynnn007

File: src/agent/src/rpc.rs
Comment (Xynnn007) [inline src/agent/src/rpc.rs:127]: Great. Let me use this socket path in CDH.

File: src/agent/src/rpc.rs
Comment (Xynnn007) [inline src/agent/src/rpc.rs:1920]: Can we combine these sentences into one `feaeture` flag? s.t.
```
    #[cfg(feature = "sealed-secret")]
    let client = {
        let c = ttrpc::asynchronous::Client::connect(CDH_ADDR)?;
        Some(sealed_secret_ttrpc_async::SealedSecretServiceClient::new(c))
    };
```

File: src/agent/src/rpc.rs
Comment (Xynnn007) [inline src/agent/src/rpc.rs:792]: I'm not sure about this:

as there are functions `unseal_secret_async()`, `remove_if_sock_exist()`, etc are related to sealed secret feature. Is it possible to put them into a separate file or a `mod` block. In this way we might only need one `#[cfg(feature = "sealed-secret")]` flag to enable the file or the `mod` block?

File: src/agent/src/rpc.rs
Comment (Xynnn007) [inline src/agent/src/rpc.rs:157]: In future there might be more APIs to be called by the client. We might need to impl all the RPCs in one Client, s.t. `CdhClient` or something else

**Review by Xynnn007 — COMMENTED** [review]
Thanks @LindaYu17 ! I'll impl CDH side using the same API definition soon.

File: src/agent/src/rpc.rs
Comment (LindaYu17) [reply src/agent/src/rpc.rs]: fixed, thanks.

File: src/agent/src/rpc.rs
Comment (LindaYu17) [reply src/agent/src/rpc.rs]: move related code to src/agent/src/cdh.rs

File: src/agent/src/rpc.rs
Comment (LindaYu17) [reply src/agent/src/rpc.rs]: wrap SealedSecretServiceClient in to CDHClient in cdh.rs, thanks.

File: src/agent/src/rpc.rs
Comment (LindaYu17) [reply src/agent/src/rpc.rs]: fixed, thanks

File: src/agent/src/rpc.rs
Comment (LindaYu17) [reply src/agent/src/rpc.rs]: thanks, and I think it is better isolate the sealed secret code with a feature

File: src/agent/src/rpc.rs
Comment (Xynnn007) [inline src/agent/src/rpc.rs:1883]: Only a little personal views :)

I found that the `ttrpc::asynchronous::Client::connect(CDH_ADDR)?` and `CDH_ADDR` const here are not needed to be exposed in `rpc.rs`. How about defining a `CDHClient::new()`, and wrap initialization of the ttrpc client code into that function? Then we can make the feature codes and non-feature codes more distinguishable.

**Review by Xynnn007 — APPROVED** [review]
LGTM. Thanks!

File: src/agent/src/rpc.rs
Comment (LindaYu17) [reply src/agent/src/rpc.rs]: fixed

**Comment by LindaYu17** [comment]
add a feature "confidential-data-hub" to  control cdh.rs, and a feature "sealed-secret" depends on "confidential-data-hub" to enable sealed secret support in kata-agent

**Comment by LindaYu17** [comment]
sealed secret as file is supported as well

**Review by fitzthum — COMMENTED** [review]
Nice work. Minor comments

File: src/agent/src/cdh.rs
Comment (Xynnn007) [reply src/agent/src/cdh.rs:50]: This value is widely used in
ocicrypt-rs: https://github.com/confidential-containers/guest-components/blob/main/ocicrypt-rs/src/keywrap/keyprovider.rs#L157
image-rs: https://github.com/confidential-containers/guest-components/blob/main/image-rs/src/resource/kbs/ttrpc.rs#L48

Comment (LindaYu17) [reply src/agent/src/cdh.rs]: Thank you @fitzthum @Xynnn007 
I just define a "const SEALED_SECRET_TIMEOUT: i64 = 50 * 1000 * 1000 * 1000;" for it, is it ok?

File: src/agent/src/rpc.rs
Comment (LindaYu17) [reply src/agent/src/rpc.rs]: fixed, combined the codes under only one feature flag, thanks.

**Comment by yoheiueda** [comment]
@LindaYu17
> sealed secret as file is supported as well

Do you have any plan to support dynamic secret data updates?

Data in a Secret Volume can be updated after container is created as described here.

https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod
> When a volume contains data from a Secret, and that Secret is updated, Kubernetes tracks this and updates the data in the volume, using an eventually-consistent approach.

In the case of peer pods, @bpradipt implemented the propagation mechanism for updated secret data (#6342). Kata shim monitors updates of ConfigMap/Secret volumes, and uses `CopyFIle` to send updated data to kata-agent when it detects any updates.

In the case of sealed secrets, I think kata agent can use a similar approach. That is, kata agent monitors updates of original secret volumes, and calls `unseal_file` when it detects any updates. WDYT?

**Comment by Xynnn007** [comment]
> In the case of peer pods, @bpradipt implemented the propagation mechanism for updated secret data (#6342). Kata shim monitors updates of ConfigMap/Secret volumes, and uses `CopyFIle` to send updated data to kata-agent when it detects any updates.
> 
> In the case of sealed secrets, I think kata agent can use a similar approach. That is, kata agent monitors updates of original secret volumes, and calls `unseal_file` when it detects any updates. WDYT?

Just a comment: any update to the secrets inside the guest could change the status of the Tee, thus updation should also be measured or filtered by a policy that has been measured during the launch time.

File: src/agent/src/cdh.rs
Comment (arronwy) [inline src/agent/src/cdh.rs:120]: remove unwrap()

File: src/agent/src/cdh.rs
Comment (LindaYu17) [reply src/agent/src/cdh.rs]: fixed, thanks

File: src/agent/src/cdh.rs
Comment (LindaYu17) [reply src/agent/src/cdh.rs]: fixed, thanks

File: src/agent/src/cdh.rs
Comment (LindaYu17) [reply src/agent/src/cdh.rs]: fixed, thanks

**Comment by LindaYu17** [comment]
> @LindaYu17
> 
> > sealed secret as file is supported as well
> 
> Do you have any plan to support dynamic secret data updates?
> 
> Data in a Secret Volume can be updated after container is created as described here.
> 
> https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod
> 
> > When a volume contains data from a Secret, and that Secret is updated, Kubernetes tracks this and updates the data in the volume, using an eventually-consistent approach.
> 
> In the case of peer pods, @bpradipt implemented the propagation mechanism for updated secret data (#6342). Kata shim monitors updates of ConfigMap/Secret volumes, and uses `CopyFIle` to send updated data to kata-agent when it detects any updates.
> 
> In the case of sealed secrets, I think kata agent can use a similar approach. That is, kata agent monitors updates of original secret volumes, and calls `unseal_file` when it detects any updates. WDYT?

Thank you @yoheiueda, really good point! We will consider to support sealed secret update. And as @Xynnn007 said, we need "agent policy feature" support as well.

**Review by arronwy — APPROVED** [review]
Thanks @LindaYu17 LGTM!

File: src/runtime/config/configuration-qemu-snp.toml.in
Comment (fitzthum) [inline src/runtime/config/configuration-qemu-snp.toml.in:712]: btw, maybe it's time for us to get rid of this placeholder section? seems like it could be very confusing as we start to implement more features 

wdyt @arronwy @stevenhorsman ?

File: src/runtime/virtcontainers/kata_agent.go
Comment (fitzthum) [inline src/runtime/virtcontainers/kata_agent.go:915]: Is this going to interfere with standard secrets that are not sealed?

File: src/runtime/config/configuration-qemu-snp.toml.in
Comment (stevenhorsman) [reply src/runtime/config/configuration-qemu-snp.toml.in:712]: Yeah, that image section pre-dates my involvement in the project, so is in `main`, not just `CCv0`, but should definitely be reviewed as part of the image pull -> main contribution. We definitely use the `service_offload` value and I think we also use `image_request_timeout`, but I don't think the provision has ever been used, or makes much sense, so I'd expect it to go

File: src/agent/src/cdh.rs
Comment (c3d) [inline src/agent/src/cdh.rs:45]: Please add quotes around "sealed.", otherwise a user would have trouble parsing this error message.

What about a message like `stripping "sealed." prefix failed` ?

File: src/agent/src/cdh.rs
Comment (c3d) [inline src/agent/src/cdh.rs:191]: I think the message is meant to be `expected a unix:// socket address` or `socket address does not have the expected format` or something like that.
(The message as written suggests you do _not_ expect a socket address, when in fact you do)/

File: src/runtime/config/configuration-qemu-se.toml.in
Comment (c3d) [inline src/runtime/config/configuration-qemu-se.toml.in:649]:
```suggestion
# If true, sealed secret feature will be enabled
# (default: false)
```

File: src/runtime/config/configuration-qemu-sev.toml.in
Comment (c3d) [inline src/runtime/config/configuration-qemu-sev.toml.in:653]:
```suggestion
# If true, sealed secret feature will be enabled
# (default: false)
```

File: src/runtime/config/configuration-qemu-snp.toml.in
Comment (c3d) [inline src/runtime/config/configuration-qemu-snp.toml.in:678]:
```suggestion
# If true, sealed secret feature will be enabled
# (default: false)
```

File: src/runtime/config/configuration-qemu-tdx.toml.in
Comment (c3d) [inline src/runtime/config/configuration-qemu-tdx.toml.in:666]:
```suggestion
# If true, sealed secret feature will be enabled
# (default: false)
```

**Review by c3d — COMMENTED** [review]
Overall, this looks pretty good.

A few comments about error messages, that should be more user-friendly if possible, and avoid internal jargon / implementation details. For each possible error, try to think about a way to formulate it that would help a user getting it fix the problem.

File: src/runtime/virtcontainers/kata_agent.go
Comment (fitzthum) [reply src/runtime/virtcontainers/kata_agent.go:915]: I think there are valid workloads that have some secrets that are privileged and need to be sealed and others that are not.

File: src/runtime/config/configuration-qemu-snp.toml.in
Comment (arronwy) [reply src/runtime/config/configuration-qemu-snp.toml.in:713]: yes, fully agree, it is confusing to new user since runtime stack is not using them.

File: src/agent/src/cdh.rs
Comment (Lu-Biao) [reply src/agent/src/cdh.rs:20]: Addded cdh document link at the header of this file
```
// Confidential Data Hub client wrapper.
// Confidential Data Hub is a service running inside guest to provide resource related APIs.
// https://github.com/confidential-containers/guest-components/tree/main/confidential-data-hub
```

File: src/agent/src/cdh.rs
Comment (Lu-Biao) [reply src/agent/src/cdh.rs:201]: If you remove 'as',  will run into compilation error:
```
error[E0308]: mismatched types
   --> src/cdh.rs:202:86
    |
202 |             let ss_service = sealed_secret_ttrpc_async::create_sealed_secret_service(ss);
    |                              ------------------------------------------------------- ^^ expected `Arc<Box<...>>`, found `Arc<Box<TestService>>`
    |                              |
    |                              arguments to this function are incorrect
    |
    = note: expected struct `std::sync::Arc<std::boxed::Box<(dyn protocols::sealed_secret_ttrpc_async::SealedSecretService + std::marker::Send + Sync + 'static)>>`
               found struct `std::sync::Arc<std::boxed::Box<TestService>>`
note: function defined here
   --> /home/lu/co/kata-image/kata-containers/src/libs/protocols/src/sealed_secret_ttrpc_async.rs:61:8
    |
61  | pub fn create_sealed_secret_service(service: Arc<Box<dyn SealedSecretService + Send + Sync>>) -> HashMap<String, ::ttrpc::r#as...
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0308`.
error: could not compile `kata-agent` (bin "kata-agent" test) due to previous error
l
```

File: src/agent/src/cdh.rs
Comment (Lu-Biao) [reply src/agent/src/cdh.rs]: fixed, thanks.

File: src/runtime/config/configuration-remote.toml.in
Comment (Lu-Biao) [reply src/runtime/config/configuration-remote.toml.in]: fixed, thanks.

**Comment by fidencio** [comment]
@LindaYu17, @Lu-Biao, I've fixed static-checks issues, and also avoided https://github.com/kata-containers/kata-containers/issues/7793 to happen here as well. I've also removed the depends-on from this side, and added it to the tests one, and finally forced pushed to @LindaYu17's CCv0 branch.

Let's see what the tests, both here and in the tests repo side, will tell us.

File: src/agent/src/cdh.rs
Comment (stevenhorsman) [inline src/agent/src/cdh.rs:121]: nit: could we replace these lines by using the convenience function, so something like:
```
let contents = fs::read_to_string(target_path)?;
```
?

File: src/agent/src/cdh.rs
Comment (stevenhorsman) [inline src/agent/src/cdh.rs:131]: nit: could we replace this with the convenience function like:
```
fs::write(unsealed_filename, unsealed_value.plaintext)
```

**Review by stevenhorsman — COMMENTED** [review]
The code looks ok to my non-expert eyes, but I've left a few comments and suggestions. I've also kicked off the tests to see how they get on.

File: src/agent/src/cdh.rs
Comment (LindaYu17) [reply src/agent/src/cdh.rs]: fixed, thanks

File: src/agent/src/cdh.rs
Comment (LindaYu17) [reply src/agent/src/cdh.rs]: fixed

File: src/agent/src/cdh.rs
Comment (LindaYu17) [reply src/agent/src/cdh.rs]: fixed, thanks

File: src/agent/src/cdh.rs
Comment (LindaYu17) [reply src/agent/src/cdh.rs]: fixed, thanks

File: src/agent/src/main.rs
Comment (LindaYu17) [reply src/agent/src/main.rs]: fixed, thanks

File: src/agent/src/main.rs
Comment (LindaYu17) [reply src/agent/src/main.rs]: fixed, thanks

File: src/agent/src/cdh.rs
Comment (stevenhorsman) [inline src/agent/src/cdh.rs:87]: nit:
```suggestion
                "sealed source path {:?} does not exist", sealed_source_path
```

**Comment by fidencio** [comment]
@LindaYu17, could you please add commit messages to each one of the PRs?  The https://github.com/kata-containers/kata-containers/actions/runs/6094755771/job/16549455954?pr=7570 tests are failing because of that.

Once that's fixed we can have this one merged.

File: src/agent/src/cdh.rs
Comment (LindaYu17) [reply src/agent/src/cdh.rs]: fixed, thanks

**Comment by LindaYu17** [comment]
> @LindaYu17, could you please add commit messages to each one of the PRs? The https://github.com/kata-containers/kata-containers/actions/runs/6094755771/job/16549455954?pr=7570 tests are failing because of that.
> 
> Once that's fixed we can have this one merged.

fixed, thanks.

**Comment by fidencio** [comment]
I propose we merge this one.

s390x is failing, but it passed before changing the commit messages: http://jenkins.katacontainers.io/job/kata-containers-CCv0-ubuntu-20.04-s390x-PR/850/

The failure is not related, just dnf being funny and not being able to download rpms.

**Comment by fidencio** [comment]
Got the green light to merge from @BbolroC! :-)

**Pull request merged**
